charms/contextual_status.py

```diff
@@ -55,22 +63,41 @@ def context(unit):
         yield
     finally:
         contexts.pop()
-        statuses = context['blocked'] + context['waiting']
+        statuses = context["blocked"] + context["waiting"]
         log.info(f"Status context closed with: {statuses}")
-        unit.status = statuses[0] if statuses else ActiveStatus()
+        unit.status = statuses[0] if statuses else exit_status
+
+
+class ReconcilerError(Exception):
+    """
+    Raised by the on_error context when the charm translates
+    a known/expected error into a Waiting or Blocked Status.
+
+    on_error allows the charm developer to escape the reconcile
+    loop by raising expected exceptions, which will be translated
+    into charm status conditions.
+    """
+
+    pass
 
 
 @contextmanager
-def on_error(status):
-    """ Context for emitting status on error.
+def on_error(status: StatusBase, *status_exceptions: Type[Exception]):
+    """Context for emitting status on error.
 
     If an exception occurs within the on_error context body, then add the
     specified status to the status context. This can be used as a function
     decorator to emit Blocked or Waiting status on error with less try/except
     boilerplate.
```

Comment:
docs update

```suggestion
    If an exception occurs within the on_error context body, then add the
    specified status to the status context. This can be used as a function
    decorator to emit Blocked or Waiting status on error with less try/except
    boilerplate.
    
    By default, on_error catches all exceptions, but it can be tuned to only
    catch specific exception types passed in status_exceptions.  When tuned,
    the status is only added to the context on the passed exceptions, and other
    exceptions won't be caught
```